tst/unit/test_pararrays.cpp

```diff
@@ -0,0 +1,365 @@
+          using policy = Kokkos::MDRangePolicy<Kokkos::Rank<2>>;
+          Kokkos::parallel_reduce(policy({0,0}, {N2,N1}),
+                                  KOKKOS_LAMBDA(const int j, const int i,
+                                                int& update) {
+                                    update += v2d(j,i);
+                                  },
+                                  sum_device);
+          REQUIRE( sum_host == sum_device );
+        }
+      }
+      THEN("slicing is possible") {
+        // auto b = a.SliceD(std::make_pair(1,3),3);
+        // auto b = a.SliceD<3>(std::make_pair(1,3));
+        auto b = a.SliceD<3>(1,2); // indx,nvar
+        AND_THEN("slices have correct values.") {
+          int total_errors = 1; // != 0
+          using policy = Kokkos::MDRangePolicy<Kokkos::Rank<3>>;
+          Kokkos::parallel_reduce(policy({0,0,0}, {2,N2,N1}),
+                                  KOKKOS_LAMBDA(const int k,
+                                                const int j,
+                                                const int i,
+                                                int& update) {
+                                    update += (b(k,j,i) == a(k+1,j,i)) ? 0 : 1;
+                                  },
+                                  total_errors);
+          REQUIRE( total_errors == 0 );
+        }
+      }
+    }
+  }
+}
+
+TEST_CASE("Time simple stencil operations") {
+  SECTION("1d range") {
+    std::cout << "1d range:" << std::endl;
+    profile_wrapper_3d(parthenon::loop_pattern_flatrange_tag);
+  }
+  /* // skip this output for now, since times are comparable
+  SECTION("md range") {
+    std::cout << "md range:" << std::endl;
+    profile_wrapper_3d(parthenon::loop_pattern_mdrange_tag);
+  }
+  SECTION("tpttr") {
+    std::cout << "tpttr range:" << std::endl;
+    profile_wrapper_3d(parthenon::loop_pattern_tpttr_tag);
+  }
+  SECTION("tpttrtvr") {
+    std::cout << "tpttrvr range:" << std::endl;
+    profile_wrapper_3d(parthenon::loop_pattern_tpttrtvr_tag);
+  }
+  */
+#ifndef KOKKOS_ENABLE_CUDA
+  /*
+  SECTION("tptvr") {
+    std::cout << "tptvr range:" << std::endl;
+    profile_wrapper_3d(parthenon::loop_pattern_tptvr_tag);
+  }
+  */
+  SECTION("simdfor") {
+    std::cout << "simd range:" << std::endl;
+    profile_wrapper_3d(parthenon::loop_pattern_simdfor_tag);
+  }
+#endif
+}
+
+TEST_CASE("Check registry pressure","[ParArrayND]") {
+  auto exec_space = DevSpace();
+  Kokkos::Timer timer;
+
+  // https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution
+  std::random_device
+    rd; // Will be used to obtain a seed for the random number engine
+  std::mt19937 gen(rd()); // Standard mersenne_twister_engine seeded with rd()
+  std::uniform_real_distribution<Real> dis(-1.0, 1.0);
+
+  // view of views. See:
+  // https://github.com/kokkos/kokkos/wiki/View
+  using arrays_t = Kokkos::View<ParArrayND<Real>*,UVMSpace>;
+  using views_t = Kokkos::View<ParArray3D<Real>*,UVMSpace>;
```

Comment:
I'm concerned that we're doing a timing test on an array of views in `UVMSpace`. It's not quite representative of a kernel with many Views coming in as parameters since it doesn't consume the same memory space. The kernel timing might also be dominated by the UVM memory access, I'm not sure.

Also, an aside that we might all already know, but in the production code we'll want no UVM anywhere since it add unnecessary GPU-CPU synchronization.

Reply:
So the `Kokkos::View`s  and `ParArrayND`s themselves are not in `UVMSpace`. They're in `DefaultExecutionSpace`. The only thing that's in `UVMSpace` is the outer view. That's required for doing views of views, because the metadata for each view has to be available both on host and device. [(See here for more info.)](https://github.com/kokkos/kokkos/wiki/View#623-can-i-make-a-view-of-views) That was the only way I could figure out to get an arbitrary-sized collection of views into a kernel. I tried an array, but `nvcc` wasn't having any of that. 

Given that the `UVMSpace` is only for the metadata and pointers, not the data itself, do you think that should be causing the issues you're describing?

Reply:
The access to `UVMSpace` itself is slow, and it happens in every thread, even though the data itself is pretty small. Performance will would also depend on whether the UVM data is prefetched to the device before executing.

According to  `nvvp`,  the kernel is limited by memory bandwidth, although it's still very compute heavy. Unfortunately, `nvvp` runs into memory errors when I try to do the `UVM` profiling. I've seen this happen before on `Kokkos` projects built with `cmake`. 

I can't really tell if `UVMSpace` explains everything we're seeing. We should still avoid it though since it introduces a lot of device-host synchronization and it will make working with concurrent streams more difficult.